lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
```

Comment:
Should we validate that child tables themselves are _not_ partitioned tables? Currently, if we are in a sub-partitioning scenario, the call to `safe_add_concurrent_index` below would raise an error and we'd be left in a state with an invalid index on the parent table

Reply:
Yes, I think so.

Actually, that could be checked in `safe_add_concurrent_index` itself, which would be even more broadly useful.

I also think it might be worth exposing this block as a public method (`def children_of_partition_parent_table(table)`?); I've definitely need to use something similar before.

Reply:
So the `CREATE INDEX ... ON ONLY` happens first before we add the indexes on child tables, so if we add the validation to `safe_add_concurrent_index` we'd still be left with an invalid index on the parent table. I think maybe it makes to have that validation in both this method and `safe_add_concurrent_index`. I'll start working on that

> I also think it might be worth exposing this block as a public method (def children_of_partition_parent_table(table)?

Yeah that makes sense, but I think I'll open a separate PR